server/cluster/unsafe_recovery_controller_test.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"bytes"
+	"context"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raft_serverpb"
+	"github.com/tikv/pd/pkg/mock/mockid"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/kv"
+)
+
+var _ = Suite(&testUnsafeRecoverSuite{})
+
+type testUnsafeRecoverSuite struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (s *testUnsafeRecoverSuite) TearDownTest(c *C) {
+	s.cancel()
+}
+
+func (s *testUnsafeRecoverSuite) SetUpTest(c *C) {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+}
+
+func (s *testUnsafeRecoverSuite) TestOneHealthyRegion(c *C) {
+	_, opt, _ := newTestScheduleConfig()
+	cluster := newTestRaftCluster(s.ctx, mockid.NewIDAllocator(), opt, core.NewStorage(kv.NewMemoryKV()), core.NewBasicCluster())
+	recoveryController := newUnsafeRecoveryController(cluster)
+	recoveryController.failedStores = map[uint64]string{
+		3: "",
+	}
+	recoveryController.storeReports = map[uint64]*pdpb.StoreReport{
+		1: &pdpb.StoreReport{StoreId: 1, Reports: []*pdpb.PeerReport{
+			&pdpb.PeerReport{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1,
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 2},
+						Peers: []*metapb.Peer{
+							&metapb.Peer{Id: 11, StoreId: 1}, &metapb.Peer{Id: 21, StoreId: 2}, &metapb.Peer{Id: 31, StoreId: 3}}}}},
+		}},
+		2: &pdpb.StoreReport{StoreId: 2, Reports: []*pdpb.PeerReport{
+			&pdpb.PeerReport{
+				RaftState: &raft_serverpb.RaftLocalState{LastIndex: 10},
+				RegionState: &raft_serverpb.RegionLocalState{
+					Region: &metapb.Region{
+						Id:          1,
+						RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 1},
```

Comment:
epoch is not the same but the range is the same, it couldn't happen.

Reply:
Done